File: apps/api/src/modules/pdp/reason.ts
import type { MissingLink, ModelIR, ReasonResult, ReasonStep } from "@lazyfga/shared";
import { gateway } from "../../openfga";
import { getCurrentVersion } from "../model/model.repo";

const MAX_DEPTH = 8;

/** explain이 의존하는 OpenFGA 연산(테스트 시 주입 가능). */
export interface ReasonDeps {
  check(
    input: { user: string; relation: string; object: string; context?: Record<string, unknown> },
    opts?: { authorizationModelId?: string },
  ): Promise<{ allowed: boolean }>;
  read(input: {
    user?: string;
    relation?: string;
    object?: string;
  }): Promise<{ tuples: { user: string; relation: string; object: string }[] }>;
}

const gatewayDeps: ReasonDeps = {
  check: (i, o) => gateway.check(i, o),
  read: (i) => gateway.read(i),
};

/** 결정과 reason이 같은 모델 버전을 쓰도록 evaluate가 핀을 전달한다. */
export interface ReasonPin {
  decision: boolean;
  authorizationModelId: string;
  ir: ModelIR;
}

interface ExplainCtx {
  authorizationModelId: string;
  ir: ModelIR;
  context?: Record<string, unknown>;
  deps: ReasonDeps;
  /** (permission|object) 방문 집합: cycle·재수렴 그래프에서 재방문/조합폭발 방지. */
  visited: Set<string>;
}

function splitObject(object: string): { type: string; id: string } | null {
  const i = object.indexOf(":");
  if (i < 0) return null;
  return { type: object.slice(0, i), id: object.slice(i + 1) };
}

/**
 * (object, role) tuple로 직접/그룹 경유를 판별. Check는 true였는데 직접도 확인된 그룹도
 * 아니면(와일드카드/비-member userset/페이지 밖 등) incomplete=true로 정직하게 표기.
 */
async function classifyRoleStep(
  user: string,
  role: string,
  object: string,
  onType: string,
  ctx: ExplainCtx,
): Promise<{ step: ReasonStep; incomplete: boolean }> {
  const { tuples } = await ctx.deps.read({ object, relation: role });
  if (tuples.some((t) => t.user === user)) {
    return { step: { via: "role", role, on: onType, direct: true }, incomplete: false };
  }
  for (const t of tuples) {
    const hash = t.user.indexOf("#");
    if (hash > 0 && t.user.slice(hash + 1) === "member") {
      const groupObject = t.user.slice(0, hash); // 예: team:eng
      const { allowed } = await ctx.deps.check(
        { user, relation: "member", object: groupObject, context: ctx.context },
        { authorizationModelId: ctx.authorizationModelId },
      );
      if (allowed) {
        return {
          step: {
            via: "role",
            role,
            on: onType,
            direct: false,
            group: splitObject(groupObject)?.type ?? groupObject,
            groupObject, // 인스턴스 수준 설명용 구체 id(예: team:eng)
          },
          incomplete: false,
        };
      }
    }
  }
  return { step: { via: "role", role, on: onType, direct: false }, incomplete: true };
}

interface WitnessResult {
  found: boolean;
  path?: ReasonStep[];
  /** 깊이/페이지 한도 또는 role-step 분류 실패로 설명이 불완전한지. */
  truncated: boolean;
}

/** 존재 증명(bounded): role 직접/그룹 → parent 상속 재귀 중 최초 성립 경로. */
async function findWitness(
  user: string,
  permission: string,
  object: string,
  ctx: ExplainCtx,
  depth: number,
): Promise<WitnessResult> {
  if (depth > MAX_DEPTH) return { found: false, truncated: true };
  const visitKey = `${permission}|${object}`;
  if (ctx.visited.has(visitKey)) return { found: false, truncated: false }; // cycle/재수렴 가드
  ctx.visited.add(visitKey);
  const parsed = splitObject(object);
  if (!parsed) return { found: false, truncated: false };
  const resource = ctx.ir.resources.find((r) => r.name === parsed.type);
  const perm = resource?.permissions.find((p) => p.name === permission);
  if (!perm) return { found: false, truncated: false };

  for (const role of perm.grantedByRoles) {
    const { allowed } = await ctx.deps.check(
      { user, relation: role, object, context: ctx.context },
      { authorizationModelId: ctx.authorizationModelId },
    );
    if (allowed) {
      const { step, incomplete } = await classifyRoleStep(user, role, object, parsed.type, ctx);
      return { found: true, path: [step], truncated: incomplete };
    }
  }

  // 형제 parent들을 끝까지 탐색: 한 가지가 깊이 초과여도 다른 깨끗한 witness를 놓치지 않는다.
  // read는 store-level tuple 목록이라 모델 버전과 무관(Check만 모델 핀이 필요).
  let sawTruncation = false;
  for (const rel of perm.inheritFromParents) {
    const { tuples } = await ctx.deps.read({ object, relation: rel });
    for (const t of tuples) {
      const child = await findWitness(user, permission, t.user, ctx, depth + 1);
      if (child.found && child.path) {
        const parentType = splitObject(t.user)?.type ?? rel;
        return {
          found: true,
          path: [
            { via: "parent", relation: rel, parent: parentType, parentObject: t.user },
            ...child.path,
          ],
          truncated: child.truncated,
        };
      }
      if (child.truncated) sawTruncation = true;
    }
  }
  return { found: false, truncated: sawTruncation };
}

function describePath(user: string, permission: string, object: string, path: ReasonStep[]): string {
  const parts = path.map((s) =>
    s.via === "role"
      ? `role ${s.role}${s.direct ? " (direct)" : s.groupObject ? ` (via ${s.groupObject} membership)` : ""}`
      : `inherited via ${s.relation} from ${s.parentObject ?? s.parent}`,
  );
  return `${user} can ${permission} ${object}: ${parts.join(" → ")}`;
}

function describeMissing(object: string, links: MissingLink[]): string {
  const bits = links.map((l) =>
    l.kind === "role" ? `one of [${l.anyOf.join(", ")}] on ${object}` : `${l.needs} via parent (${l.relation})`,
  );
  return `denied: needs ${bits.length ? bits.join(", or ") : "a grant that does not exist in the model"}`;
}

function denyLinks(ir: ModelIR, onType: string, permission: string): MissingLink[] {
  const perm = ir.resources.find((r) => r.name === onType)?.permissions.find((p) => p.name === permission);
  const links: MissingLink[] = [];
  if (perm) {
    if (perm.grantedByRoles.length > 0) links.push({ kind: "role", anyOf: perm.grantedByRoles, on: onType });
    for (const rel of perm.inheritFromParents) links.push({ kind: "parent", relation: rel, needs: `can_${permission}` });
  }
  return links;
}

/**
 * 결정에 대한 사람이 읽는 reason. 허용=witnessing path, 거부=missing links(비대칭).
 * pin이 주어지면 그 모델 버전/decision을 그대로 사용(evaluate와 동일 버전 보장, DB 재조회 없음).
 */
export async function explain(
  user: string,
  permission: string,
  object: string,
  context?: Record<string, unknown>,
  pin?: ReasonPin,
  deps: ReasonDeps = gatewayDeps,
): Promise<ReasonResult> {
  let authorizationModelId: string;
  let ir: ModelIR;
  let decision: boolean;

  if (pin) {
    ({ authorizationModelId, ir, decision } = pin);
  } else {
    const current = await getCurrentVersion();
    if (!current) return { decision: false, text: "model not published" };
    authorizationModelId = current.authorizationModelId;
    ir = current.irJson;
    const { allowed } = await deps.check(
      { user, relation: `can_${permission}`, object, context },
      { authorizationModelId },
    );
    decision = allowed;
  }

  const ctx: ExplainCtx = { authorizationModelId, ir, context, deps, visited: new Set() };
  const onType = splitObject(object)?.type ?? object;

  if (decision) {
    const w = await findWitness(user, permission, object, ctx, 0);
    if (w.found && w.path) {
      const text = describePath(user, permission, object, w.path);
      return w.truncated
        ? { decision: true, path: w.path, truncated: true, text: `${text} (partial)` }
        : { decision: true, path: w.path, text };
    }
    return {
      decision: true,
      truncated: true,
      text: `allowed via can_${permission} (path reconstruction incomplete)`,
    };
  }

  const missingLinks = denyLinks(ir, onType, permission);
  return { decision: false, missingLinks, text: describeMissing(object, missingLinks) };
}

File: apps/api/internal/modules/pdp/reason.go
// Package pdp는 AuthZEN 평가와 reason 엔진을 담당한다(TS modules/pdp 포팅).
// 단일 질문 템플릿으로 OpenFGA Check 1회를 수행하고, 요청 시 witnessing path(허용) 또는
// missing links(거부)를 재구성한다. 결정과 reason은 같은 모델 버전(pin)을 사용한다.
package pdp

import (
	"context"
	"strings"

	"github.com/antegral-constellation/lazyfga/api/internal/contract"
	"github.com/antegral-constellation/lazyfga/api/internal/openfga"
)

// maxDepth는 reason 그래프 탐색의 최대 상속 깊이다.
const maxDepth = 8

// Gateway는 evaluate/reason이 필요로 하는 OpenFGA 연산이다(테스트에서 fake 주입).
type Gateway interface {
	Check(ctx context.Context, in openfga.CheckInput, opts ...openfga.CheckOption) (bool, error)
	Read(ctx context.Context, in openfga.ReadInput) ([]openfga.ReadTuple, error)
}

// reasonPin은 결정과 reason이 동일 모델 버전/decision을 쓰도록 evaluate가 넘기는 핀이다.
type reasonPin struct {
	decision             bool
	authorizationModelID string
	ir                   *contract.ModelIR
}

// explainCtx는 한 explain 호출의 공유 상태다. visited는 cycle·재수렴 가드((permission|object)).
type explainCtx struct {
	authorizationModelID string
	ir                   *contract.ModelIR
	context              map[string]any
	deps                 Gateway
	visited              map[string]bool
}

func splitObject(object string) (typ, id string, ok bool) {
	i := strings.Index(object, ":")
	if i < 0 {
		return "", "", false
	}
	return object[:i], object[i+1:], true
}

// classifyRoleStep은 (object, role) tuple로 직접/그룹 경유를 판별한다. Check는 true였는데 직접도
// 확인된 그룹도 아니면(와일드카드/비-member userset/페이지 밖 등) incomplete=true로 정직하게 표기한다.
func classifyRoleStep(ctx context.Context, cx *explainCtx, user, role, object, onType string) (contract.ReasonStep, bool, error) {
	tuples, err := cx.deps.Read(ctx, openfga.ReadInput{Object: &object, Relation: &role})
	if err != nil {
		return contract.ReasonStep{}, false, err
	}
	for _, t := range tuples {
		if t.User == user {
			return contract.ReasonStep{Via: "role", Role: role, On: onType, Direct: true}, false, nil
		}
	}
	for _, t := range tuples {
		hash := strings.Index(t.User, "#")
		if hash > 0 && t.User[hash+1:] == "member" {
			groupObject := t.User[:hash] // 예: team:eng
			allowed, err := cx.deps.Check(ctx,
				openfga.CheckInput{User: user, Relation: "member", Object: groupObject, Context: cx.context},
				openfga.WithCheckAuthorizationModelID(cx.authorizationModelID))
			if err != nil {
				return contract.ReasonStep{}, false, err
			}
			if allowed {
				groupType := groupObject
				if gt, _, ok := splitObject(groupObject); ok {
					groupType = gt
				}
				go2 := groupObject
				return contract.ReasonStep{Via: "role", Role: role, On: onType, Direct: false, Group: &groupType, GroupObject: &go2}, false, nil
			}
		}
	}
	return contract.ReasonStep{Via: "role", Role: role, On: onType, Direct: false}, true, nil
}

type witnessResult struct {
	found     bool
	path      []contract.ReasonStep
	truncated bool
}

// findWitness는 role 직접/그룹 → parent 상속 재귀 중 최초 성립 경로를 bounded 탐색한다.
func findWitness(ctx context.Context, cx *explainCtx, user, permission, object string, depth int) (witnessResult, error) {
	if depth > maxDepth {
		return witnessResult{found: false, truncated: true}, nil
	}
	visitKey := permission + "|" + object
	if cx.visited[visitKey] {
		return witnessResult{found: false, truncated: false}, nil // cycle/재수렴 가드.
	}
	cx.visited[visitKey] = true

	typ, _, ok := splitObject(object)
	if !ok {
		return witnessResult{found: false, truncated: false}, nil
	}
	perm := findPermission(cx.ir, typ, permission)
	if perm == nil {
		return witnessResult{found: false, truncated: false}, nil
	}

	for _, role := range perm.GrantedByRoles {
		allowed, err := cx.deps.Check(ctx,
			openfga.CheckInput{User: user, Relation: role, Object: object, Context: cx.context},
			openfga.WithCheckAuthorizationModelID(cx.authorizationModelID))
		if err != nil {
			return witnessResult{}, err
		}
		if allowed {
			step, incomplete, err := classifyRoleStep(ctx, cx, user, role, object, typ)
			if err != nil {
				return witnessResult{}, err
			}
			return witnessResult{found: true, path: []contract.ReasonStep{step}, truncated: incomplete}, nil
		}
	}

	// 형제 parent들을 끝까지 탐색: 한 가지가 깊이 초과여도 다른 깨끗한 witness를 놓치지 않는다.
	sawTruncation := false
	for _, rel := range perm.InheritFromParents {
		rel := rel
		tuples, err := cx.deps.Read(ctx, openfga.ReadInput{Object: &object, Relation: &rel})
		if err != nil {
			return witnessResult{}, err
		}
		for _, t := range tuples {
			child, err := findWitness(ctx, cx, user, permission, t.User, depth+1)
			if err != nil {
				return witnessResult{}, err
			}
			if child.found && child.path != nil {
				parentType := rel
				if pt, _, ok := splitObject(t.User); ok {
					parentType = pt
				}
				parentObject := t.User
				path := append([]contract.ReasonStep{{Via: "parent", Relation: rel, Parent: parentType, ParentObject: &parentObject}}, child.path...)
				return witnessResult{found: true, path: path, truncated: child.truncated}, nil
			}
			if child.truncated {
				sawTruncation = true
			}
		}
	}
	return witnessResult{found: false, truncated: sawTruncation}, nil
}

func findPermission(ir *contract.ModelIR, typ, permission string) *contract.Permission {
	for i := range ir.Resources {
		if ir.Resources[i].Name != typ {
			continue
		}
		for j := range ir.Resources[i].Permissions {
			if ir.Resources[i].Permissions[j].Name == permission {
				return &ir.Resources[i].Permissions[j]
			}
		}
		return nil
	}
	return nil
}

func describePath(user, permission, object string, path []contract.ReasonStep) string {
	parts := make([]string, len(path))
	for i, s := range path {
		if s.Via == "role" {
			suffix := ""
			if s.Direct {
				suffix = " (direct)"
			} else if s.GroupObject != nil {
				suffix = " (via " + *s.GroupObject + " membership)"
			}
			parts[i] = "role " + s.Role + suffix
		} else {
			from := s.Parent
			if s.ParentObject != nil {
				from = *s.ParentObject
			}
			parts[i] = "inherited via " + s.Relation + " from " + from
		}
	}
	return user + " can " + permission + " " + object + ": " + strings.Join(parts, " → ")
}

func describeMissing(object string, links []contract.MissingLink) string {
	bits := make([]string, len(links))
	for i, l := range links {
		if l.Kind == "role" {
			bits[i] = "one of [" + strings.Join(l.AnyOf, ", ") + "] on " + object
		} else {
			bits[i] = l.Needs + " via parent (" + l.Relation + ")"
		}
	}
	joined := "a grant that does not exist in the model"
	if len(bits) > 0 {
		joined = strings.Join(bits, ", or ")
	}
	return "denied: needs " + joined
}

func denyLinks(ir *contract.ModelIR, onType, permission string) []contract.MissingLink {
	links := []contract.MissingLink{}
	perm := findPermission(ir, onType, permission)
	if perm != nil {
		if len(perm.GrantedByRoles) > 0 {
			links = append(links, contract.MissingLink{Kind: "role", AnyOf: perm.GrantedByRoles, On: onType})
		}
		for _, rel := range perm.InheritFromParents {
			links = append(links, contract.MissingLink{Kind: "parent", Relation: rel, Needs: "can_" + permission})
		}
	}
	return links
}

// explain은 결정에 대한 사람이 읽는 reason을 만든다. 허용=witnessing path, 거부=missing links(비대칭).
// pin의 모델 버전/decision을 그대로 사용한다(evaluate와 동일 버전 보장, DB 재조회 없음).
func explain(ctx context.Context, deps Gateway, pin reasonPin, user, permission, object string, reqCtx map[string]any) (contract.ReasonResult, error) {
	cx := &explainCtx{
		authorizationModelID: pin.authorizationModelID,
		ir:                   pin.ir,
		context:              reqCtx,
		deps:                 deps,
		visited:              map[string]bool{},
	}
	onType := object
	if t, _, ok := splitObject(object); ok {
		onType = t
	}

	if pin.decision {
		w, err := findWitness(ctx, cx, user, permission, object, 0)
		if err != nil {
			return contract.ReasonResult{}, err
		}
		if w.found && w.path != nil {
			text := describePath(user, permission, object, w.path)
			if w.truncated {
				tr := true
				return contract.ReasonResult{Decision: true, Path: w.path, Truncated: &tr, Text: text + " (partial)"}, nil
			}
			return contract.ReasonResult{Decision: true, Path: w.path, Text: text}, nil
		}
		tr := true
		return contract.ReasonResult{Decision: true, Truncated: &tr, Text: "allowed via can_" + permission + " (path reconstruction incomplete)"}, nil
	}

	missing := denyLinks(pin.ir, onType, permission)
	return contract.ReasonResult{Decision: false, MissingLinks: missing, Text: describeMissing(object, missing)}, nil
}
